### DISCUSSION TITLE
File picker is really slow on huge projects

### DISCUSSION THREAD

**Discussion by liquidev** [helix-editor/helix#10995] [General]
### Summary

I just tried editing the Unreal Engine source code with Helix, and ` f`ing into an arbitrary file from my project. Unfortunately it seems like Helix indexes the entire source directory recursively from scratch every time you open the file picker, which results in unacceptably slow performance in a huge codebase like Unreal Engine's. (As of typing this I typed in the name of the file I'd like to open, and the file picker still hasn't indexed my file.)

This is also compounded by Helix indexing directories that are listed in the `.ignore` file. It seems like `.ignore` just results in files matching any of the globs being hidden from the view, rather than preventing recursion.

### Reproduction Steps

1. Open a _compiled_ Unreal Engine in Helix (since it has way more files than just the cloned source code)
2. ` f`
3. Search for a file such as `EdGraphPin.cpp`

### Helix log

Irrelevant since this is strictly a performance issue.

### Platform

Windows

### Terminal Emulator

Windows Terminal

### Installation Method

winget

### Helix Version

helix 24.3 (2cadec0b)

**Comment by the-mikedavis** [comment]
The log isn't irrelevant, this line might be insightful: https://github.com/helix-editor/helix/blob/31bcde360c6dbe9931ec3d43964172d0e9d964f7/helix-term/src/ui/mod.rs#L218

Can you provide a simpler reproduction case - a different similarly large project? Cloning unreal engine is convoluted

**Comment by pascalkuthe** [comment]
Our transversal is currently not parallel because we sort entries alphabetically.

Windows has super slow IO. Anytime I see people having issues with IO speed like this it's Windows (if you are on WSL its even worse still) on Linux/macos it's never an issue for projects of practical size (I work on massive legacy codebases). 

We can likely offer an option to disable the sorting and allow parallel transversal. It's not too useful since it's only a secondary tie breaker as soon as you enter a query. 

Caching the filesystem does not make sense. For small directories it's not necessary and for large directories the memory consumption is prohibitive (you wouldn't want to cache $HOME or /). 

And .ignore does prevent entering directories if a glob matches a directory

**Reply by liquidev** [reply]
With the amount of files in UE I suppose sorting entries above a certain number of results would not make too much sense anyways, so maybe it's a good solution. I wonder how much faster parallel directory traversal is on Windows, though.

> And .ignore does prevent entering directories if a glob matches a directory

I do wonder then why sometimes the `(running)` text disappears for a while, and then resumes - could it be because it's hitting lots of ignored files within non-ignored subdirectories?

**Reply by kirawi** [reply]
I've created an issue to track sorting as a config option: https://github.com/helix-editor/helix/issues/11021

**Comment by markstos** [comment]
> Can you provide a simpler reproduction case - a different similarly large project?

The Linux kernel is often used when people need a huge project to test with. 

I just tried untar'ing this tarball of 85,000 files and then opening `hx` in the directory:

https://git.kernel.org/torvalds/t/linux-6.10-rc2.tar.gz

It opened instantly. On Linux.

**Comment by RoloEdits** [comment]
@liquidev One thing you can try is the new dev drive (built on top of ReFS) windows 11 comes with (you didn't specify windows version, so assuming 11). I've noticed various I/O improvements myself. Your milage may vary.

link: https://learn.microsoft.com/en-us/windows/dev-drive/

**Reply by liquidev** [reply]
Unfortunately my work computer runs Windows 10 for now so I'm unable to use Dev Drive. I assume it would improve the performance somewhat, I wonder just how much exactly.

I tried opening the Unreal Engine Git repository on Linux though, and although it has way less files than a full Unreal Engine game, the entire directory was indexed in just over a second. So that's way better already, but unfortunately "just use Linux" is not a solution that's viable for work.

**Reply by markstos** [reply]
WSL2 is supported in Windows 10. I've heard it performs well. https://learn.microsoft.com/en-us/windows/wsl/compare-versions

**Comment by jamal** [comment]
Hello, apologies for reviving a very old conversation. I'm on a similar boat as @liquidev working on a very large UE project on Windows. That being said, I'm on Win11 and using ReFS for the project and still find it somewhat slow. It looks like there are two issues:
1. When I open the file picker (`<space>f f`) it takes Helix hangs for 3 seconds or so before it shows it. Though, I hadn't noticed this before so this may be something new. I recently compiled the steel system branch and could totally be related to changes there.
2. Scanning through the project takes a while, so finding files in the project can take a minute or so.

A lot of the time I try to just use file explorer instead and navigate to the file I want, but it would be much more convenient to be able to open arbitrary files (specially when I don't know where they are).

I'm curious if there is anything that can be done to improve this or if it's still an open issue. Thanks!
